### PR TITLE
Add new int4 quantization to dense_vector

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new quantization mechanism for HNSW and flat indices. Here we add `int4` quantization via the `int4_hnsw` and `int4_flat` index types. This quantization methodology further reduces the memory required for fast HNSW, meaning that the memory required is 8x smaller than with regular float32 values. 

8x reduction means that 1M 1024 dimension vectors goes from requiring 3.8GB to 477MB.

Recall continues to stay steady, there is some reduction that is recoverable via slightly oversampling and reranking. For example over 500k CohereV3 vectors, only 5 extra vectors are required to be gathered to achieve over 0.98 recall in a brute-force scenario.

recall

### CODE REVIEW

**Comment by benwtrent** [comment]
As for query speed, I rally on my local macbook (so, load driver & server on the same box, but numbers were pretty consistent). 

Some highlights:

 - Store size does increase 9.54%
 - Refresh time is slightly slower
 - index-concurrent-with-searches is 10% faster
 - force-merge is almost 30% faster
 - QPS is between 30-50% faster (with one measurement saying its near 2x better qps)

```
------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------

|                                                        Metric |                                         Task |       Baseline |        Contender |          Diff |   Unit |   Diff % |
|--------------------------------------------------------------:|---------------------------------------------:|---------------:|-----------------:|--------------:|-------:|---------:|
|                    Cumulative indexing time of primary shards |                                              |   10.9729      |     10.9267      |      -0.04623 |    min |   -0.42% |
|             Min cumulative indexing time across primary shard |                                              |    5.47465     |      5.42623     |      -0.04842 |    min |   -0.88% |
|          Median cumulative indexing time across primary shard |                                              |    5.48645     |      5.46333     |      -0.02312 |    min |   -0.42% |
|             Max cumulative indexing time across primary shard |                                              |    5.49825     |      5.50043     |       0.00218 |    min |   +0.04% |
|           Cumulative indexing throttle time of primary shards |                                              |    0           |      0           |       0       |    min |    0.00% |
|    Min cumulative indexing throttle time across primary shard |                                              |    0           |      0           |       0       |    min |    0.00% |
| Median cumulative indexing throttle time across primary shard |                                              |    0           |      0           |       0       |    min |    0.00% |
|    Max cumulative indexing throttle time across primary shard |                                              |    0           |      0           |       0       |    min |    0.00% |
|                       Cumulative merge time of primary shards |                                              |   23.6884      |     17.3226      |      -6.3658  |    min |  -26.87% |
|                      Cumulative merge count of primary shards |                                              |   18           |     14           |      -4       |        |  -22.22% |
|                Min cumulative merge time across primary shard |                                              |   11.6643      |      8.63782     |      -3.02647 |    min |  -25.95% |
|             Median cumulative merge time across primary shard |                                              |   11.8442      |      8.66131     |      -3.1829  |    min |  -26.87% |
|                Max cumulative merge time across primary shard |                                              |   12.0241      |      8.6848      |      -3.33933 |    min |  -27.77% |
|              Cumulative merge throttle time of primary shards |                                              |    0.62935     |      0.570283    |      -0.05907 |    min |   -9.39% |
|       Min cumulative merge throttle time across primary shard |                                              |    0.302333    |      0.237283    |      -0.06505 |    min |  -21.52% |
|    Median cumulative merge throttle time across primary shard |                                              |    0.314675    |      0.285142    |      -0.02953 |    min |   -9.39% |
|       Max cumulative merge throttle time across primary shard |                                              |    0.327017    |      0.333       |       0.00598 |    min |   +1.83% |
|                     Cumulative refresh time of primary shards |                                              |    0.15305     |      0.316683    |       0.16363 |    min | +106.91% |
|                    Cumulative refresh count of primary shards |                                              |  188           |    157           |     -31       |        |  -16.49% |
|              Min cumulative refresh time across primary shard |                                              |    0.06985     |      0.156433    |       0.08658 |    min | +123.96% |
|           Median cumulative refresh time across primary shard |                                              |    0.076525    |      0.158342    |       0.08182 |    min | +106.91% |
|              Max cumulative refresh time across primary shard |                                              |    0.0832      |      0.16025     |       0.07705 |    min |  +92.61% |
|                       Cumulative flush time of primary shards |                                              |    0.15435     |      0.21985     |       0.0655  |    min |  +42.44% |
|                      Cumulative flush count of primary shards |                                              |   14           |     14           |       0       |        |    0.00% |
|                Min cumulative flush time across primary shard |                                              |    0.0759833   |      0.1067      |       0.03072 |    min |  +40.43% |
|             Median cumulative flush time across primary shard |                                              |    0.077175    |      0.109925    |       0.03275 |    min |  +42.44% |
|                Max cumulative flush time across primary shard |                                              |    0.0783667   |      0.11315     |       0.03478 |    min |  +44.39% |
|                                       Total Young Gen GC time |                                              |    1.523       |      2.599       |       1.076   |      s |  +70.65% |
|                                      Total Young Gen GC count |                                              |  111           |    186           |      75       |        |  +67.57% |
|                                         Total Old Gen GC time |                                              |    0           |      0           |       0       |      s |    0.00% |
|                                        Total Old Gen GC count |                                              |    0           |      0           |       0       |        |    0.00% |
|                                                    Store size |                                              |    1.27557     |      1.39729     |       0.12173 |     GB |   +9.54% |
|                                                 Translog size |                                              |    1.02445e-07 |      1.02445e-07 |       0       |     GB |    0.00% |
|                                        Heap used for segments |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                      Heap used for doc values |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                           Heap used for terms |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                           Heap used for norms |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                          Heap used for points |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                   Heap used for stored fields |                                              |    0           |      0           |       0       |     MB |    0.00% |
|                                                 Segment count |                                              |    2           |      2           |       0       |        |    0.00% |
|                                   Total Ingest Pipeline count |                                              |    0           |      0           |       0       |        |    0.00% |
|                                    Total Ingest Pipeline time |                                              |    0           |      0           |       0       |     ms |    0.00% |
|                                  Total Ingest Pipeline failed |                                              |    0           |      0           |       0       |        |    0.00% |
|                                                Min Throughput |                                 index-append | 6074.56        |   6151.82        |      77.2575  | docs/s |   +1.27% |
|                                               Mean Throughput |                                 index-append | 6669.01        |   6779.89        |     110.875   | docs/s |   +1.66% |
|                                             Median Throughput |                                 index-append | 6440.64        |   6526.34        |      85.7032  | docs/s |   +1.33% |
|                                                Max Throughput |                                 index-append | 8625.35        |   8792.29        |     166.94    | docs/s |   +1.94% |
|                                       50th percentile latency |                                 index-append |  896.696       |    882.216       |     -14.4802  |     ms |   -1.61% |
|                                       90th percentile latency |                                 index-append | 1030.53        |   1037.87        |       7.34429 |     ms |   +0.71% |
|                                       99th percentile latency |                                 index-append | 1343.57        |   1196.64        |    -146.933   |     ms |  -10.94% |
|                                      100th percentile latency |                                 index-append | 1509.43        |   1740.53        |     231.101   |     ms |  +15.31% |
|                                  50th percentile service time |                                 index-append |  896.696       |    882.216       |     -14.4802  |     ms |   -1.61% |
|                                  90th percentile service time |                                 index-append | 1030.53        |   1037.87        |       7.34429 |     ms |   +0.71% |
|                                  99th percentile service time |                                 index-append | 1343.57        |   1196.64        |    -146.933   |     ms |  -10.94% |
|                                 100th percentile service time |                                 index-append | 1509.43        |   1740.53        |     231.101   |     ms |  +15.31% |
|                                                    error rate |                                 index-append |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                          refresh-after-index |    1.59311     |      1.3654      |      -0.22771 |  ops/s |  -14.29% |
|                                               Mean Throughput |                          refresh-after-index |    1.59311     |      1.3654      |      -0.22771 |  ops/s |  -14.29% |
|                                             Median Throughput |                          refresh-after-index |    1.59311     |      1.3654      |      -0.22771 |  ops/s |  -14.29% |
|                                                Max Throughput |                          refresh-after-index |    1.59311     |      1.3654      |      -0.22771 |  ops/s |  -14.29% |
|                                      100th percentile latency |                          refresh-after-index |  625.111       |    729.643       |     104.532   |     ms |  +16.72% |
|                                 100th percentile service time |                          refresh-after-index |  625.111       |    729.643       |     104.532   |     ms |  +16.72% |
|                                                    error rate |                          refresh-after-index |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |        index-update-concurrent-with-searches | 7567.59        |   8302.08        |     734.487   | docs/s |   +9.71% |
|                                               Mean Throughput |        index-update-concurrent-with-searches | 7725.55        |   8577.89        |     852.344   | docs/s |  +11.03% |
|                                             Median Throughput |        index-update-concurrent-with-searches | 7673.32        |   8580.12        |     906.801   | docs/s |  +11.82% |
|                                                Max Throughput |        index-update-concurrent-with-searches | 8159.34        |   8710.41        |     551.074   | docs/s |   +6.75% |
|                                       50th percentile latency |        index-update-concurrent-with-searches |  659.994       |    552.274       |    -107.72    |     ms |  -16.32% |
|                                       90th percentile latency |        index-update-concurrent-with-searches |  739.125       |    619.681       |    -119.444   |     ms |  -16.16% |
|                                      100th percentile latency |        index-update-concurrent-with-searches |  759.621       |    742.573       |     -17.048   |     ms |   -2.24% |
|                                  50th percentile service time |        index-update-concurrent-with-searches |  659.994       |    552.274       |    -107.72    |     ms |  -16.32% |
|                                  90th percentile service time |        index-update-concurrent-with-searches |  739.125       |    619.681       |    -119.444   |     ms |  -16.16% |
|                                 100th percentile service time |        index-update-concurrent-with-searches |  759.621       |    742.573       |     -17.048   |     ms |   -2.24% |
|                                                    error rate |        index-update-concurrent-with-searches |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput | knn-search-100-1000-concurrent-with-indexing |   57.4021      |     62.6         |       5.1979  |  ops/s |   +9.06% |
|                                               Mean Throughput | knn-search-100-1000-concurrent-with-indexing |   60.5556      |     63.9116      |       3.35594 |  ops/s |   +5.54% |
|                                             Median Throughput | knn-search-100-1000-concurrent-with-indexing |   60.5377      |     63.9285      |       3.39082 |  ops/s |   +5.60% |
|                                                Max Throughput | knn-search-100-1000-concurrent-with-indexing |   61.8569      |     65.4762      |       3.61925 |  ops/s |   +5.85% |
|                                       50th percentile latency | knn-search-100-1000-concurrent-with-indexing |   14.4036      |     13.7783      |      -0.62525 |     ms |   -4.34% |
|                                       90th percentile latency | knn-search-100-1000-concurrent-with-indexing |   17.8694      |     16.8782      |      -0.99121 |     ms |   -5.55% |
|                                       99th percentile latency | knn-search-100-1000-concurrent-with-indexing |   24.6098      |     23.1164      |      -1.4934  |     ms |   -6.07% |
|                                     99.9th percentile latency | knn-search-100-1000-concurrent-with-indexing |   36.8631      |     31.01        |      -5.85309 |     ms |  -15.88% |
|                                      100th percentile latency | knn-search-100-1000-concurrent-with-indexing |   50.6247      |     45.4115      |      -5.21317 |     ms |  -10.30% |
|                                  50th percentile service time | knn-search-100-1000-concurrent-with-indexing |   14.4036      |     13.7783      |      -0.62525 |     ms |   -4.34% |
|                                  90th percentile service time | knn-search-100-1000-concurrent-with-indexing |   17.8694      |     16.8782      |      -0.99121 |     ms |   -5.55% |
|                                  99th percentile service time | knn-search-100-1000-concurrent-with-indexing |   24.6098      |     23.1164      |      -1.4934  |     ms |   -6.07% |
|                                99.9th percentile service time | knn-search-100-1000-concurrent-with-indexing |   36.8631      |     31.01        |      -5.85309 |     ms |  -15.88% |
|                                 100th percentile service time | knn-search-100-1000-concurrent-with-indexing |   50.6247      |     45.4115      |      -5.21317 |     ms |  -10.30% |
|                                                    error rate | knn-search-100-1000-concurrent-with-indexing |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                         refresh-after-update |  205.876       |     54.6397      |    -151.236   |  ops/s |  -73.46% |
|                                               Mean Throughput |                         refresh-after-update |  205.876       |     54.6397      |    -151.236   |  ops/s |  -73.46% |
|                                             Median Throughput |                         refresh-after-update |  205.876       |     54.6397      |    -151.236   |  ops/s |  -73.46% |
|                                                Max Throughput |                         refresh-after-update |  205.876       |     54.6397      |    -151.236   |  ops/s |  -73.46% |
|                                      100th percentile latency |                         refresh-after-update |    3.93779     |     17.1989      |      13.2611  |     ms | +336.77% |
|                                 100th percentile service time |                         refresh-after-update |    3.93779     |     17.1989      |      13.2611  |     ms | +336.77% |
|                                                    error rate |                         refresh-after-update |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |          knn-search-10-100_multiple_segments |  150.607       |    186.566       |      35.9589  |  ops/s |  +23.88% |
|                                               Mean Throughput |          knn-search-10-100_multiple_segments |  161.357       |    199.952       |      38.5951  |  ops/s |  +23.92% |
|                                             Median Throughput |          knn-search-10-100_multiple_segments |  161.351       |    200.696       |      39.3445  |  ops/s |  +24.38% |
|                                                Max Throughput |          knn-search-10-100_multiple_segments |  170.101       |    210.518       |      40.417   |  ops/s |  +23.76% |
|                                       50th percentile latency |          knn-search-10-100_multiple_segments |    4.38185     |      3.43131     |      -0.95054 |     ms |  -21.69% |
|                                       90th percentile latency |          knn-search-10-100_multiple_segments |    6.08818     |      4.91559     |      -1.17259 |     ms |  -19.26% |
|                                       99th percentile latency |          knn-search-10-100_multiple_segments |    9.64134     |      6.64716     |      -2.99417 |     ms |  -31.06% |
|                                     99.9th percentile latency |          knn-search-10-100_multiple_segments |   15.8832      |     10.8362      |      -5.04702 |     ms |  -31.78% |
|                                      100th percentile latency |          knn-search-10-100_multiple_segments |   16.4266      |     18.3915      |       1.96492 |     ms |  +11.96% |
|                                  50th percentile service time |          knn-search-10-100_multiple_segments |    4.38185     |      3.43131     |      -0.95054 |     ms |  -21.69% |
|                                  90th percentile service time |          knn-search-10-100_multiple_segments |    6.08818     |      4.91559     |      -1.17259 |     ms |  -19.26% |
|                                  99th percentile service time |          knn-search-10-100_multiple_segments |    9.64134     |      6.64716     |      -2.99417 |     ms |  -31.06% |
|                                99.9th percentile service time |          knn-search-10-100_multiple_segments |   15.8832      |     10.8362      |      -5.04702 |     ms |  -31.78% |
|                                 100th percentile service time |          knn-search-10-100_multiple_segments |   16.4266      |     18.3915      |       1.96492 |     ms |  +11.96% |
|                                                    error rate |          knn-search-10-100_multiple_segments |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |        knn-search-100-1000_multiple_segments |   82.2396      |    103.967       |      21.7273  |  ops/s |  +26.42% |
|                                               Mean Throughput |        knn-search-100-1000_multiple_segments |   84.1761      |    106.432       |      22.2557  |  ops/s |  +26.44% |
|                                             Median Throughput |        knn-search-100-1000_multiple_segments |   84.237       |    105.745       |      21.5081  |  ops/s |  +25.53% |
|                                                Max Throughput |        knn-search-100-1000_multiple_segments |   85.5745      |    110.613       |      25.0386  |  ops/s |  +29.26% |
|                                       50th percentile latency |        knn-search-100-1000_multiple_segments |   10.395       |      8.62781     |      -1.76717 |     ms |  -17.00% |
|                                       90th percentile latency |        knn-search-100-1000_multiple_segments |   12.3902      |     10.2701      |      -2.1201  |     ms |  -17.11% |
|                                       99th percentile latency |        knn-search-100-1000_multiple_segments |   14.1842      |     11.3321      |      -2.85205 |     ms |  -20.11% |
|                                     99.9th percentile latency |        knn-search-100-1000_multiple_segments |   20.3963      |     17.7157      |      -2.68063 |     ms |  -13.14% |
|                                      100th percentile latency |        knn-search-100-1000_multiple_segments |   23.3523      |     22.8707      |      -0.4815  |     ms |   -2.06% |
|                                  50th percentile service time |        knn-search-100-1000_multiple_segments |   10.395       |      8.62781     |      -1.76717 |     ms |  -17.00% |
|                                  90th percentile service time |        knn-search-100-1000_multiple_segments |   12.3902      |     10.2701      |      -2.1201  |     ms |  -17.11% |
|                                  99th percentile service time |        knn-search-100-1000_multiple_segments |   14.1842      |     11.3321      |      -2.85205 |     ms |  -20.11% |
|                                99.9th percentile service time |        knn-search-100-1000_multiple_segments |   20.3963      |     17.7157      |      -2.68063 |     ms |  -13.14% |
|                                 100th percentile service time |        knn-search-100-1000_multiple_segments |   23.3523      |     22.8707      |      -0.4815  |     ms |   -2.06% |
|                                                    error rate |        knn-search-100-1000_multiple_segments |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                                  force-merge |    0.000826405 |      0.00115818  |       0.00033 |  ops/s |  +40.15% |
|                                               Mean Throughput |                                  force-merge |    0.000826405 |      0.00115818  |       0.00033 |  ops/s |  +40.15% |
|                                             Median Throughput |                                  force-merge |    0.000826405 |      0.00115818  |       0.00033 |  ops/s |  +40.15% |
|                                                Max Throughput |                                  force-merge |    0.000826405 |      0.00115818  |       0.00033 |  ops/s |  +40.15% |
|                                      100th percentile latency |                                  force-merge |    1.21006e+06 | 863422           | -346636       |     ms |  -28.65% |
|                                 100th percentile service time |                                  force-merge |    1.21006e+06 | 863422           | -346636       |     ms |  -28.65% |
|                                                    error rate |                                  force-merge |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                            knn-search-10-100 |  147.037       |    184.179       |      37.1425  |  ops/s |  +25.26% |
|                                               Mean Throughput |                            knn-search-10-100 |  152.854       |    230.357       |      77.5027  |  ops/s |  +50.70% |
|                                             Median Throughput |                            knn-search-10-100 |  150.136       |    205.873       |      55.7371  |  ops/s |  +37.12% |
|                                                Max Throughput |                            knn-search-10-100 |  173.418       |    328.923       |     155.505   |  ops/s |  +89.67% |
|                                       50th percentile latency |                            knn-search-10-100 |    5.048       |      4.2056      |      -0.8424  |     ms |  -16.69% |
|                                       90th percentile latency |                            knn-search-10-100 |    8.02059     |      7.69148     |      -0.32911 |     ms |   -4.10% |
|                                       99th percentile latency |                            knn-search-10-100 |   11.8411      |      9.69997     |      -2.14109 |     ms |  -18.08% |
|                                     99.9th percentile latency |                            knn-search-10-100 |   13.7798      |     10.9792      |      -2.80051 |     ms |  -20.32% |
|                                      100th percentile latency |                            knn-search-10-100 |   15.6253      |     15.7229      |       0.09762 |     ms |   +0.62% |
|                                  50th percentile service time |                            knn-search-10-100 |    5.048       |      4.2056      |      -0.8424  |     ms |  -16.69% |
|                                  90th percentile service time |                            knn-search-10-100 |    8.02059     |      7.69148     |      -0.32911 |     ms |   -4.10% |
|                                  99th percentile service time |                            knn-search-10-100 |   11.8411      |      9.69997     |      -2.14109 |     ms |  -18.08% |
|                                99.9th percentile service time |                            knn-search-10-100 |   13.7798      |     10.9792      |      -2.80051 |     ms |  -20.32% |
|                                 100th percentile service time |                            knn-search-10-100 |   15.6253      |     15.7229      |       0.09762 |     ms |   +0.62% |
|                                                    error rate |                            knn-search-10-100 |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                          knn-search-100-1000 |  109.275       |    131.164       |      21.8883  |  ops/s |  +20.03% |
|                                               Mean Throughput |                          knn-search-100-1000 |  112.425       |    136.821       |      24.3956  |  ops/s |  +21.70% |
|                                             Median Throughput |                          knn-search-100-1000 |  112.831       |    137.673       |      24.8412  |  ops/s |  +22.02% |
|                                                Max Throughput |                          knn-search-100-1000 |  113.307       |    138.191       |      24.8832  |  ops/s |  +21.96% |
|                                       50th percentile latency |                          knn-search-100-1000 |    6.86156     |      5.8025      |      -1.05906 |     ms |  -15.43% |
|                                       90th percentile latency |                          knn-search-100-1000 |    9.0475      |      7.37713     |      -1.67037 |     ms |  -18.46% |
|                                       99th percentile latency |                          knn-search-100-1000 |   13.1252      |      8.68652     |      -4.43871 |     ms |  -33.82% |
|                                     99.9th percentile latency |                          knn-search-100-1000 |   17.5741      |     12.1126      |      -5.46148 |     ms |  -31.08% |
|                                      100th percentile latency |                          knn-search-100-1000 |   17.9156      |     12.1784      |      -5.73721 |     ms |  -32.02% |
|                                  50th percentile service time |                          knn-search-100-1000 |    6.86156     |      5.8025      |      -1.05906 |     ms |  -15.43% |
|                                  90th percentile service time |                          knn-search-100-1000 |    9.0475      |      7.37713     |      -1.67037 |     ms |  -18.46% |
|                                  99th percentile service time |                          knn-search-100-1000 |   13.1252      |      8.68652     |      -4.43871 |     ms |  -33.82% |
|                                99.9th percentile service time |                          knn-search-100-1000 |   17.5741      |     12.1126      |      -5.46148 |     ms |  -31.08% |
|                                 100th percentile service time |                          knn-search-100-1000 |   17.9156      |     12.1784      |      -5.73721 |     ms |  -32.02% |
|                                                    error rate |                          knn-search-100-1000 |    0           |      0           |       0       |      % |    0.00% |
|                                                Min Throughput |                          
```

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 1]
```diff
@@ -142,6 +152,27 @@ PUT my-byte-quantized-index
...
+[source,console]
+--------------------------------------------------
+PUT my-byte-quantized-index
+{
+  "mappings": {
+    "properties": {
+      "my_vector": {
+        "type": "dense_vector",
+        "dims": 4,
+        "index": true,
```
Comment (jimczi) [inline docs/reference/mapping/types/dense-vector.asciidoc:166]: nit: that's the default so maybe remove it

**Review by jimczi — APPROVED** [review]
LGTM, nice set of tests!

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 2]
```diff
@@ -1032,12 +1076,148 @@ public int hashCode() {
...
+
+        @Override
+        void validateDimension(int dim) {
+            if (dim % 2 != 0) {
+                throw new IllegalArgumentException("int4_hnsw only supports even dimensions; provided=" + dim);
+            }
+        }
+    }
+
+    static class Int4FlatIndexOptions extends IndexOptions {
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1154]: Nit: Would it make sense for this class to extend `Int8FlatIndexOption`, or extract a common base class?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 3]
```diff
@@ -1032,12 +1076,148 @@ public int hashCode() {
         }
     }
 
-    private static class Int8HnswIndexOptions extends IndexOptions {
+    static class Int4HnswIndexOptions extends IndexOptions {
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1079]: Nit: Would it make sense for this class to extend `Int8HnswIndexOptions`, or extract a common base class?

**Review by carlosdelest — APPROVED** [review]
LGTM 👍

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 4]
```diff
@@ -1032,12 +1076,147 @@ public int hashCode() {
...
+
+        @Override
+        boolean supportsElementType(ElementType elementType) {
+            return elementType != ElementType.BYTE;
+        }
+
+        @Override
+        boolean updatableTo(IndexOptions update) {
+            // TODO: add support for updating from flat, hnsw, and int8_hnsw and updating params
+            return Objects.equals(this, update);
```
Comment (tteofili) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1203]: this would be nice to add. we might also do it in a separate PR.

**Review by tteofili — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/codec/vectors/ES813Int8FlatVectorFormat.java [hunk 5]
```diff
@@ -37,16 +37,15 @@ public class ES813Int8FlatVectorFormat extends KnnVectorsFormat {
...
     public ES813Int8FlatVectorFormat() {
-        this(null);
+        this(null, 7, false);
     }
 
     /**
      * Sole constructor
      */
-    public ES813Int8FlatVectorFormat(Float confidenceInterval) {
+    public ES813Int8FlatVectorFormat(Float confidenceInterval, int bits, boolean compress) {
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/ES813Int8FlatVectorFormat.java:46]: Are all combinations of allowed `bits` and  `compress` allowed/functional? If not, maybe we could add some validation/extract the `compress` value based on `bits`?

**Review by pmpailis — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 4 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1203]: @tteofili I was thinking a separate PR due to the additional testing & code churn.

**Comment by benwtrent** [comment]
@elasticmachine update branch
